src/main/java/org/logstash/beats/BeatsParser.java

```diff
@@ -0,0 +1,230 @@
+package org.logstash.beats;
+
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.InflaterInputStream;
+
+
+public class BeatsParser extends ByteToMessageDecoder {
+    private static final int CHUNK_SIZE = 1024;
+    private final static Logger logger = LogManager.getLogger(Server.class.getName());
+
+    private Batch batch = new Batch();
+
+    private enum States {
+        READ_HEADER,
+        READ_FRAME_TYPE,
+        READ_WINDOW_SIZE,
+        READ_JSON_HEADER,
+        READ_COMPRESSED_FRAME_HEADER,
+        READ_COMPRESSED_FRAME,
+        READ_JSON,
+        READ_DATA_FIELDS,
+    }
+
+    private States currentState = States.READ_HEADER;
+    private long requiredBytes = 0;
+    private int sequence = 0;
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if(!hasEnoughBytes(in)) {
+            return;
+        }
+
+        switch (currentState) {
+            case READ_HEADER: {
+                logger.debug("Running: READ_HEADER");
+
+                byte currentVersion = in.readByte();
+
+                if(Protocol.isVersion2(currentVersion)) {
+                    logger.debug("Frame version 2 detected");
+                    batch.setProtocol(Protocol.VERSION_2);
+                } else {
+                    logger.debug("Frame version 1 detected");
+                    batch.setProtocol(Protocol.VERSION_1);
+                }
+
+                transition(States.READ_FRAME_TYPE, 1);
```

Comment:
minor cleanup, you could put the size of the types into the enum, so they are not magic numbers hanging around...